persistence/nosql/persistence/metastore/src/main/java/org/apache/polaris/persistence/nosql/metastore/mutation/MutationAttempt.java

```diff
@@ -0,0 +1,593 @@
+import static org.apache.polaris.persistence.nosql.metastore.ContentIdentifier.indexKeyToIdentifierBuilder;
+import static org.apache.polaris.persistence.nosql.metastore.indexaccess.IndexUtils.hasChildren;
+import static org.apache.polaris.persistence.nosql.metastore.mutation.MutationResults.newMutableMutationResults;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+import org.apache.polaris.core.entity.PolarisBaseEntity;
+import org.apache.polaris.core.entity.PolarisEntityCore;
+import org.apache.polaris.core.entity.PolarisEntityType;
+import org.apache.polaris.core.entity.PolarisPrincipalSecrets;
+import org.apache.polaris.core.persistence.dao.entity.BaseResult;
+import org.apache.polaris.core.storage.StorageLocation;
+import org.apache.polaris.persistence.nosql.api.commit.CommitterState;
+import org.apache.polaris.persistence.nosql.api.index.IndexKey;
+import org.apache.polaris.persistence.nosql.api.index.UpdatableIndex;
+import org.apache.polaris.persistence.nosql.api.obj.ObjRef;
+import org.apache.polaris.persistence.nosql.coretypes.ContainerObj;
+import org.apache.polaris.persistence.nosql.coretypes.ObjBase;
+import org.apache.polaris.persistence.nosql.coretypes.catalog.EntityIdSet;
+import org.apache.polaris.persistence.nosql.coretypes.changes.Change;
+import org.apache.polaris.persistence.nosql.coretypes.changes.ChangeAdd;
+import org.apache.polaris.persistence.nosql.coretypes.changes.ChangeRemove;
+import org.apache.polaris.persistence.nosql.coretypes.changes.ChangeRename;
+import org.apache.polaris.persistence.nosql.coretypes.changes.ChangeUpdate;
+import org.apache.polaris.persistence.nosql.coretypes.content.PolicyObj;
+import org.apache.polaris.persistence.nosql.coretypes.realm.PolicyMappingsObj;
+import org.apache.polaris.persistence.nosql.metastore.ContentIdentifier;
+import org.apache.polaris.persistence.nosql.metastore.committers.ChangeResult;
+import org.apache.polaris.persistence.nosql.metastore.indexaccess.MemoizedIndexedAccess;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public record MutationAttempt(
+    UpdateKeyForCatalogAndEntityType updateKeyForCatalogAndEntityType,
+    List<EntityUpdate> updates,
+    CommitterState<? extends ContainerObj, MutationResults> state,
+    UpdatableIndex<ObjRef> byName,
+    UpdatableIndex<IndexKey> byId,
+    UpdatableIndex<Change> changes,
+    UpdatableIndex<EntityIdSet> locations,
+    MemoizedIndexedAccess memoizedIndexedAccess) {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MutationAttempt.class);
+
+  public static ObjBase objForChangeComparison(
+      PolarisBaseEntity entity,
+      Optional<PolarisPrincipalSecrets> currentSecrets,
+      ObjBase originalObj) {
+    return mapToObj(entity, currentSecrets)
+        .updateTimestamp(originalObj.createTimestamp())
```

Comment:
Is this intentional to use `originalObj.createTimestamp` as the `updateTimestamp` for the change-comparison object? If so, definitely needs some extensive comments explaining why.

If the intent is to canonicalize `updateTimestamp` to not be relevant for semantic comparison, it might be better to set it to 0L or something to indicate intentional/obvious missing values rather than having comparison objects floating around that have the `createTimestamp` written in.

Reply:
Oops, should have been `updateTimestamp`.
Sure, this could be optimized, but the call sites would become more complicated.